documentation/src/docs/watermarking.md

```diff
@@ -0,0 +1,121 @@
+
+The time may indicate the time the data was written to the stream (ingestion time).  Time may alternatively be equated to event time (the time the event occurred).  Time can denote wall clock time, some other traditional time, or an arbitrary value, like 6.    
+
+Time has two requirements:
+
+* Time is a variable of type long.  
+* Time must increase in value (not decrease). That is, if time is 6, then a time of 5 is earlier, and a time of 7 is later.
+
+## How Pravega Writers note the time
+
+There are three ways for Pravega Writers to attach time to events.
+* Explicitly note the time  
+* Note time on transaction commit
+* Automatically note wall clock time
+
+### Explicitly note time
+
+The following writer API method notes time.
+
+`writer.noteTime(long timestamp);`
+
+You may call the method after every event write, or use it periodically to indicate  the time below which all events are written.
+
+Here is an example:       
+
+      EventStreamWriter<EventType> writer = clientFactory.createEventWriter(stream, serializer,
+                EventWriterConfig.builder().build());
+
+      //... write events ...
+
+      writer.noteTime(currentTime);
+
+### Note time on transaction commit
+
+For transactional writes, the commit call can supply the timestamp. The following Writer API method passes the time to the Controller as part of the commit. The  commit simultaneously commits the data and supplies the corresponding timestamp.
+
+`void commit(long timestamp) throws TxnFailedException;`
+
+For example:
+
+    Transaction<EventType> txn = writer.beginTxn();
+      //... write events to transaction.
+    txn.commit(txnTimestamp);
+
+### Automatically attach wall clock time
+
+The `automaticallyNoteTime` option when a  Writer is created configures the   writer to automatically attach a wall clock time notation to every event. This option essentially attaches the ingest time to each event. With this option turned on, no other calls to note the time are required.
+
+Here is the call to configure automatic timestamps on events:
+
+    EventStreamWriter<EventType> writer = clientFactory.createEventWriter(stream, serializer,
+                EventWriterConfig.builder().automaticallyNoteTime(true).build());
+
+
+## How Pravega Readers request watermark windows  
+
+Multiple Pravega Readers working in parallel each have different current locations in the stream. The Controller coordinates the Reader positions and can return a window that indicates where in the stream all of the readers are.  
+
+The following Reader API method requests the current watermark window.
+
+`TimeWindow window = reader.getTimeWindow();`
+
+The `TimeWindow` is returned as:
+
+    @Data
```

Comment:
It's best not to use lombok annotations in our public documentation as not every uses/understands lombok.